### PR TITLE
Application Crypto cleanup

### PR DESCRIPTION
**Pull request by davxy**
- Blanket implementation of `RuntimeAppPublic` trait for any type implementing `AppPublic` and for which `AppPubic::Generic` implements `RuntimePublic` (in practice implements the trait for any application-specific `Public` that can be used within the runtime).
	- Moving out this from `app_crypto!` allows to use the macro to implement application  specific crypto types that are not necessarily usable by the runtime.
	- This is required for upcoming BLS crypto key types and Sassafras keys as well

- Blanket implementation of `BoundToRuntimeAppPublic` trait for any type implementing `RuntimeAppPublic`

- Remove unused `CRYPTO_ID` from `RuntimeAppPublic`

- Documentation improvement

- Relax some trait bounds

### CODE REVIEW

**Review by michalkucharczyk — APPROVED** [review]
lgtm. Out of curiosity - why did you remove `MaybeHash`? Is this because it was not used anywhere? Or maybe it was to restrictive for some type that will be used in this context in future?

**Comment by davxy** [comment]
> lgtm. Out of curiosity - why did you remove `MaybeHash`? Is this because it was not used anywhere? Or maybe it was to restrictive for some type that will be used in this context in future?

Because I'm dumb 🤦‍♂️ 
Is required by cumulus. Restoring it

**Comment by altaua** [comment]
This needs to be rebased on current master, there have been breaking changes to the check-crate-publishing job.

File: primitives/application-crypto/src/lib.rs [hunk 1]
```diff
@@ -78,14 +79,15 @@ macro_rules! app_crypto {
...
-/// Declares Public, Pair, Signature types which are functionally equivalent to `$pair`, but are new
-/// Application-specific types whose identifier is `$key_type`.
+/// Declares `Public`, `Pair` and `Signature` types which are functionally equivalent
+/// to the corresponding types defined by `$module` but that are new application-specific
+/// types whose identifier is `$key_type`.
 ///
 /// ```rust
-/// # use sp_application_crypto::{app_crypto, wrap, ed25519, KeyTypeId};
+/// # use sp_application_crypto::{app_crypto, ed25519, KeyTypeId};
 /// // Declare a new set of crypto types using Ed25519 logic that identifies as `KeyTypeId`
```
Comment (melekes) [inline primitives/application-crypto/src/lib.rs:88]:
```suggestion
/// // Declare a new set of crypto types using ed25519 logic that identifies as `KeyTypeId`
```

**Review by melekes — COMMENTED** [review]
lgtm

**Review by skunert — APPROVED** [review]

**Pull request merged**